src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -288,20 +300,199 @@ private static ProjectItemGroupElement CreateItemGroup(Project project, string f
+        /// A simple check for some of the evaluated properties to check
+        /// if the project is package reference project or not
+        /// </summary>
+        /// <param name="project"></param>
+        /// <returns></returns>
+        internal static bool IsPackageReferenceProject(Project project)
+        {
+            return (project.GetPropertyValue(RESTORE_STYLE_TAG) != "" ||
+                    project.GetPropertyValue(PACKAGE_REFERENCE_TYPE_TAG) != "" ||
+                    project.GetPropertyValue(NUGET_STYLE_TAG) != "" ||
+                    project.GetPropertyValue(ASSETS_FILE_PATH_TAG) != "");
+        }
+
+        /// <summary>
+        /// Prepares the dictionary that maps frameworks to packages top-level
+        /// and transitive.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <param name="userInputFrameworks">A list of framework names</param>
+        /// <param name="assetsFile">Assets file for all targets and libraries</param>
+        /// <param name="transitive">Include transitive packages in the result</param>
+        /// <returns></returns>
+        internal IEnumerable<FrameworkPackages> GetResolvedVersions(
+            Project project, IEnumerable<string> userInputFrameworks,LockFile assetsFile, bool transitive)
+        {
+            var resultPackages = new List<FrameworkPackages>();
+            var parsedUserFrameworks = userInputFrameworks.Select(f => NuGetFramework.Parse(f));
+
+            IEnumerable<TargetFrameworkInformation> requestedTargetFrameworks;
+
+            if (!parsedUserFrameworks.Any())
+            {
+                requestedTargetFrameworks = assetsFile.PackageSpec.TargetFrameworks;
+            }
+            else
+            {
+                requestedTargetFrameworks = assetsFile.PackageSpec.TargetFrameworks.Where(tfm => parsedUserFrameworks.Contains(tfm.FrameworkName));
+            }
+
+            var requestedTargets = requestedTargetFrameworks.Select(tfm => assetsFile.GetTarget(tfm.FrameworkName, null));
+
+            foreach (var target in requestedTargets)
+            {
+                // Find the tfminformation corresponding to the target to
+                // get the top-level dependencies
+                var tfmInformation = requestedTargetFrameworks.Where(tfm => tfm.FrameworkName.Equals(target.TargetFramework));
+
+                //The packages for the framework that were retrieved with GetRequestedVersions
+                var frameworkDependencies = tfmInformation.First().Dependencies;
+
+                var projPackages = GetPackageReferencesPerFramework(project, "", tfmInformation.ToString());
+
+                var topLevelPackages = new List<InstalledPackageReference>();
+                var transitivePackages = new List<InstalledPackageReference>();
+
+                foreach (var library in target.Libraries)
+                {
+                    var matchingPackages = frameworkDependencies.Where(d =>
```

Comment:
use `ToList` here to convert from Enumerable to List so that we don't end up running multiple enumerations at multiple source lines.

Reply:
I don't get the issue with keeping it as an enumerable. At the end of the day this must be a single package

Reply:
With keeping Enumerable you run multiple Enumeration everytime you go through `matchingPackages` which is not performance optimized. You can also read about `List` vs `IEnumerable`